tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/Stackable.scala

```diff
@@ -0,0 +1,66 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+
+/**
+ * An object, the state of which can be saved in a stack-like manner by push and later restored by pop, similar to an
+ * SMT context. This trait is used for: SMT contexts and value caches.
+ *
+ * Some objects maintain a certain contract with the SMT solver, that is, they may add constraints to a
+ * [[at.forsyte.apalache.tla.bmcmt.smt.SolverContext]] on demand.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+trait Stackable {
```

Comment:
I'm not sure `Stackable` is a great name. Stackable suggests to me that a thing operates as a stack, but while this interface has methods `push` and `pop`, they don't have the signatures I'd expect of a stack.

iiuc, 'stack' is only showing up here because of a concept used in designing (but not actually implementing) the implementation? That is, the idea is that we have an interface for an object which can "timestamp" points in its history and then undo changes back to certain points. This is thought of as being like a stack of states, thus borrowing the terms `push` and `pop` from stacks. But the trait itself seems too have too general of a name. Maybe like "Snapshottable" would be better?

I'm also not sure `push` is a great name for the method that takes this snapshot, because it is not pushing anything. Maybe `takeSnapshot` would be clearer? And perhaps `revert` or `undo` for `pop`?

WDYT?

One other question: do you actually expect this trait to be implemented by other classes? It seems very special purpose given the last three methods. Perhaps we can do without the trait for now and just stick with the abstract class?